packages/gatsby-plugin-mdx/src/gatsby-layout-loader.ts

```diff
@@ -12,71 +12,140 @@ export interface IGatsbyLayoutLoaderOptions {
+      ecmaVersion: 2020,
+      sourceType: `module`,
+      locations: true,
+    })
+
+    const AST = tree as unknown as Program
+
+    // Throw when tree is not a Program
+    if (!AST.body && !AST.sourceType) {
+      throw new Error(
+        `Invalid AST. Parsed source code did not return a Program`
+      )
+    }
+
+    // Inject import to actual MDX file at the top of the file
+    AST.body.unshift({
+      type: `ImportDeclaration`,
+      specifiers: [
+        {
+          type: `ImportDefaultSpecifier`,
+          local: {
+            type: `Identifier`,
+            name: `GATSBY_COMPILED_MDX`,
+          },
+        },
+      ],
+      source: {
+        type: `Literal`,
+        value: mdxPath,
+      },
+    })
+
+    // Replace default export with wrapper function that injects compiled MDX as children
+    AST.body = AST.body.map(child => {
+      if (child.type !== `ExportDefaultDeclaration`) {
+        return child
+      }
+      const declaration = child.declaration as unknown as Identifier
+      if (!declaration.name) {
+        throw new Error(`Unable to determine default export name`)
+      }
+
+      const pageComponentName = declaration.name
+
+      return {
+        type: `ExportDefaultDeclaration`,
+        declaration: {
+          type: `ArrowFunctionExpression`,
+          id: null,
+          expression: true,
+          generator: false,
+          async: false,
+          params: [
+            {
+              type: `Identifier`,
+              name: `props`,
+            },
+          ],
+          body: {
+            type: `JSXElement`,
+            openingElement: {
+              type: `JSXOpeningElement`,
+              attributes: [
+                {
+                  type: `JSXSpreadAttribute`,
+                  argument: {
+                    type: `Identifier`,
+                    name: `props`,
+                  },
+                },
+              ],
+              name: {
+                type: `JSXIdentifier`,
+                name: pageComponentName,
+              },
+              selfClosing: false,
+            },
+            closingElement: {
+              type: `JSXClosingElement`,
+              name: {
+                type: `JSXIdentifier`,
+                name: pageComponentName,
+              },
+            },
+            children: [
+              {
+                type: `JSXExpressionContainer`,
+                expression: {
+                  type: `CallExpression`,
+                  callee: {
+                    type: `Identifier`,
+                    name: `GATSBY_COMPILED_MDX`,
+                  },
+                  arguments: [],
+                  optional: false,
+                },
+              },
+            ],
+          },
+        },
+      } as unknown as ExportDefaultDeclaration
+    })
+
+    // @todo what do we do with runtime, pragma and pragmaFrag options? We should still be able to support preact.
```

Comment:
Shouldn't this be handled by babel afterwards? As long as we spit out a jsx file we should be good

Reply:
Noted this down to my personal notes for the upcoming test PR